core/pkg/artifacts/saver.go

```diff
@@ -144,99 +146,157 @@ func (as *ArtifactSaver) uploadFiles(artifactID string, manifest *Manifest, mani
 			Md5:                entry.Digest,
 			ArtifactManifestID: &manifestID,
 		}
-		fileSpecs = append(fileSpecs, fileSpec)
+		fileSpecs[name] = fileSpec
 	}
 
-	// Upload in batches.
-	numInProgress, numDone := 0, 0
-	nameToScheduledTime := map[string]time.Time{}
-	taskResultsChan := make(chan TaskResult)
-	fileSpecsBatch := make([]gql.CreateArtifactFileSpecInput, 0, batchSize)
-	for numDone < len(fileSpecs) {
-		// Prepare a batch.
-		now := time.Now()
-		fileSpecsBatch = fileSpecsBatch[:0]
-		for _, fileSpec := range fileSpecs {
-			if _, ok := nameToScheduledTime[fileSpec.Name]; ok {
-				continue
-			}
-			nameToScheduledTime[fileSpec.Name] = now
-			fileSpecsBatch = append(fileSpecsBatch, fileSpec)
-			if len(fileSpecsBatch) >= batchSize {
-				break
-			}
+	var err error
+	for len(fileSpecs) > 0 {
+		numNeedUploading := len(fileSpecs)
+		if fileSpecs, err = as.processFiles(manifest, fileSpecs); err != nil {
+			return err
 		}
-		if len(fileSpecsBatch) > 0 {
-			// Fetch upload URLs.
-			response, err := gql.CreateArtifactFiles(
-				as.Ctx,
-				as.GraphqlClient,
-				fileSpecsBatch,
-				gql.ArtifactStorageLayoutV2,
+		// If more than half of the remaining files uploaded we'll keep retrying.
+		// We shouldn't ordinarily need to retry at all: our internal client handles
+		// retryable errors, and the only failure this retry loop is for is when signed
+		// urls expire before an upload was started (exceedingly rare).
+		// Still, as long as more than half of them succeed in each iteration this will
+		// eventually terminate, so we're generous with our retry policy.
+		if len(fileSpecs) > numNeedUploading/2 {
```

Comment:
It is kind of arbitrary. Keep in mind that this should be an exceedingly rare failure for small upload numbers, because all normal "retryable" errors are already retried by the underlying handler. In the example with a single file, if it fails twice that either means there's an unrecoverable error and we shouldn't retry, or that after a signed URL expired and the server requested a new lease the next signed URL also expired.

For very slow uploads this could however be too conservative. If a single batch is going to take >48 hours to upload, we actually would fail this condition because more than half of the URLs will have expired by the time we start their respective uploads; this would be an argument for adding in an additional escape for the batch taking a long time to process, as well as probably reducing the batch size in that case.